File: src/main/java/xyz/acrylicstyle/hackReport/commands/MuteCommand.java
package xyz.acrylicstyle.hackReport.commands;

import org.bukkit.Bukkit;
import org.bukkit.ChatColor;
import org.bukkit.command.Command;
import org.bukkit.command.CommandExecutor;
import org.bukkit.command.CommandSender;
import org.bukkit.entity.Player;
import util.CollectionList;
import util.ICollectionList;
import xyz.acrylicstyle.hackReport.HackReport;
import xyz.acrylicstyle.hackReport.utils.Utils;
import xyz.acrylicstyle.hackReport.utils.Webhook;
import xyz.acrylicstyle.tomeito_api.sounds.Sound;

import java.awt.*;
import java.io.IOException;
import java.util.UUID;

public class MuteCommand implements CommandExecutor {
    @Override
    public boolean onCommand(CommandSender player, Command command, String label, String[] args) {
        if (args.length == 0) {
            player.sendMessage(ChatColor.RED + "/mute <Player> " + ChatColor.GRAY + "- " + ChatColor.AQUA + "プレイヤーをミュート/ミュート解除します。");
            return true;
        }
        CollectionList<UUID> list = ICollectionList.asList(HackReport.config.getStringList("muted")).map(UUID::fromString);
        UUID uuid = IgnoreCommand.getUniqueId(args[0], player);
        if (uuid == null) return true;
        if (list.contains(uuid)) {
            list.remove(uuid);
            HackReport.config.setThenSave("muted", list.map(UUID::toString).toList());
            CollectionList<String> a = ICollectionList.asList(args);
            a.shift();
            new CollectionList<Player>(Bukkit.getOnlinePlayers()).filter(Player::isOp).forEach(p -> {
                p.playSound(p.getLocation(), Sound.BLOCK_NOTE_PLING, 100, 1);
                //p.sendMessage(ChatColor.YELLOW + player.getName() + ChatColor.GREEN + "が" + ChatColor.RED + args[0] + ChatColor.GREEN + "のミュートを解除しました。");
                Webhook webhook = Utils.getWebhook();
                if (webhook == null) return;
                new Thread(() -> {
                    webhook.addEmbed(
                            new Webhook.EmbedObject()
                                    .setTitle("`" + player.getName() + "`が`" + args[0] + "`のミュートを解除しました。")
                                    .setColor(Color.GREEN)
                                    .setDescription("理由: " + a.join(" "))
                    );
                    try {
                        webhook.execute();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }).start();
            });
            Bukkit.broadcastMessage(ChatColor.GREEN + player.getName() + "が" + args[0] + "のミュートを解除しました: " + a.join(" "));
        } else {
            list.add(uuid);
            HackReport.config.setThenSave("muted", list.map(UUID::toString).toList());
            CollectionList<String> a = ICollectionList.asList(args);
            a.shift();
            new CollectionList<Player>(Bukkit.getOnlinePlayers()).filter(Player::isOp).forEach(p -> {
                p.playSound(p.getLocation(), Sound.BLOCK_NOTE_PLING, 100, 1);
                //p.sendMessage(ChatColor.YELLOW + player.getName() + ChatColor.GREEN + "が" + ChatColor.RED + args[0] + ChatColor.GREEN + "をミュートしました。");
                Webhook webhook = Utils.getWebhook();
                if (webhook == null) return;
                new Thread(() -> {
                    webhook.addEmbed(
                            new Webhook.EmbedObject()
                                    .setTitle("`" + player.getName() + "`が`" + args[0] + "`をミュートしました。")
                                    .setColor(Color.RED)
                                    .setDescription("理由: " + a.join(" "))
                    );
                    try {
                        webhook.execute();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }).start();
            });
            Bukkit.broadcastMessage(ChatColor.GREEN + player.getName() + "が" + args[0] + "をミュートしました: " + a.join(" "));
        }
        return true;
    }
}

File: src/main/java/xyz/acrylicstyle/hackReport/commands/MuteCommand.kt
package xyz.acrylicstyle.hackReport.commands

import org.bukkit.Bukkit
import org.bukkit.ChatColor
import org.bukkit.command.Command
import org.bukkit.command.CommandExecutor
import org.bukkit.command.CommandSender
import org.bukkit.entity.Player
import util.CollectionList
import util.ICollectionList
import xyz.acrylicstyle.hackReport.struct.Mute
import xyz.acrylicstyle.hackReport.utils.ConnectionHolder.Companion.muteList
import xyz.acrylicstyle.hackReport.utils.Webhook
import xyz.acrylicstyle.joinChecker.utils.Utils
import xyz.acrylicstyle.tomeito_api.TomeitoAPI
import xyz.acrylicstyle.tomeito_api.sounds.Sound
import xyz.acrylicstyle.tomeito_api.utils.TypeUtil
import java.awt.Color
import java.util.function.Consumer

class MuteCommand : CommandExecutor {
    override fun onCommand(player: CommandSender, command: Command, label: String, args: Array<String>): Boolean {
        Thread t@ {
            if (Utils.modCheck(player)) return@t
            if (args.isEmpty()) {
                player.sendMessage("${ChatColor.RED}/mute <Player> [<time> <m/h/d>] [reason] ${ChatColor.GRAY}- ${ChatColor.AQUA}プレイヤーをミュート/ミュート解除します。")
                return@t
            }
            mute(player, args, false)
        }.start()
        return true
    }

    companion object {
        const val DAY: Long = 86_400_000
        const val HOUR: Long = 3_600_000
        const val MINUTE: Long =  60_000

        fun mute(player: CommandSender, args: Array<String>, tell: Boolean) {
            val ps = args[0]
            val arg: CollectionList<String>
            run {
                val list = ICollectionList.asList(args) // doing this in block intentionally, to prevent access to "list" from outside of this
                if (args.size > 1) {
                    if (args[1].matches("\\d+[mhd]".toRegex())) {
                        list.shift()
                        list.shift()
                        list.add(0, args[1].replace("\\d+([mhd])".toRegex(), "$1"))
                        list.add(0, args[1].replace("(\\d+)[mhd]".toRegex(), "$1"))
                        list.add(0, ps)
                    }
                }
                arg = list.clone()
            }
            val uuid = IgnoreCommand.getUniqueId(arg.shift(), player) ?: return
            val d7 = System.currentTimeMillis() + 7 * DAY
            var expiresAt = -1L
            if (arg.size >= 3) {
                if (TypeUtil.isInt(arg[0])) {
                    expiresAt = when (arg[1]) {
                        "d" -> System.currentTimeMillis() + arg[0].toLong() * DAY
                        "h" -> System.currentTimeMillis() + arg[0].toLong() * HOUR
                        "m" -> System.currentTimeMillis() + arg[0].toLong() * MINUTE
                        else -> -1
                    }
                    if (expiresAt != -1L) arg.shiftChain().shiftChain()
                }
            }
            if (expiresAt == -1L) expiresAt = d7
            val reason = arg.join(" ")
            val emptyReason = arg.isEmpty()
            if (muteList.contains(uuid, tell)) {
                muteList.remove(uuid, tell)
                TomeitoAPI.getOnlineOperators().forEach(Consumer { p: Player ->
                    p.playSound(p.location, Sound.BLOCK_NOTE_PLING, 100f, 1f)
                    if (tell) p.sendMessage("${ChatColor.GREEN}${player.name}が" + ps + "のTellミュートを解除しました" + if (emptyReason) "。" else ": $reason")
                })
                if (!tell) Bukkit.broadcastMessage("${ChatColor.GREEN}${player.name}が" + ps + "のミュートを解除しました" + if (emptyReason) "。" else ": $reason")
                Webhook.sendWebhook("`${player.name}`が`${ps}`の${if (tell) "Tell" else ""}ミュートを解除しました。", "理由: $reason", Color.GREEN)
            } else {
                if (emptyReason) {
                    player.sendMessage("${ChatColor.RED}理由を指定してください。")
                    return
                }
                muteList.add(
                    Mute(
                        uuid,
                        ps,
                        reason,
                        if (player is Player) player.uniqueId else null,
                        System.currentTimeMillis(),
                        expiresAt,
                    ),
                    tell
                )
                TomeitoAPI.getOnlineOperators().forEach(Consumer { p: Player ->
                    p.playSound(p.location, Sound.BLOCK_NOTE_PLING, 100f, 1f)
                    if (tell) p.sendMessage("${ChatColor.GREEN}${player.name}が${ps}をTellミュートしました: $reason")
                })
                if (!tell) Bukkit.broadcastMessage("${ChatColor.GREEN}${player.name}が${ps}をミュートしました: $reason")
                Webhook.sendWebhook("`${player.name}`が`${ps}`を${if (tell) "Tell" else ""}ミュートしました。", "理由: $reason", Color.RED)
            }
        }
    }
}
